driver-core/src/main/com/mongodb/internal/operation/AsyncQueryBatchCursor.java

```diff
@@ -373,6 +377,11 @@ private void handleGetMoreQueryResult(final AsyncConnection connection, final Si
         }
     }
 
+    private void logQueryResult(final QueryResult<T> result) {
+        LOGGER.debug(format("Received batch of %d documents with cursorId %d from server %s", result.getResults().size(),
+                result.getCursorId(), result.getAddress()));
```

Comment:
I don't think this is a good use of a toString method.  Note that the Javadoc also says: "The string output is not necessarily stable over time". And that's precisely the problem.  It's not uncommon for developers to change the implementation of toString when the class changes (I've often deleted toString and then used the IDEA auto-generation support to recreate it).  In contrast, we want to be very careful about the stability of anything that the driver logs.

As an alternative, I could move this to a helper method, maybe in `OperationHelper` since that's where the logger is also defined. What do you think?

Reply:
The phrase about stability is there to discourage non-human usage of the return strings. Anyway, regarding our case (the driver), if we want to keep log messages stable<sup>1</sup>, we should keep `toString` methods stable to the same extent. The only reason why we may not be able to keep `toString` consistent to the same degree we want log messages to be consistent is if we use `toString` for something else in the driver. As far as I know, we don't. Do we?

Note also that if we decide to log something additional about the state of `QueryResult` in the future, it will not matter whether we need to change its `toString` or to explicitly change the message format in the code - we will identically change the log message format as seen by a user either way. That is, again, unless the driver uses `toString` for something different than logging, nothing prevents us from using it for logging<sup>2</sup>.

> It's not uncommon for developers to change the implementation of toString when the class changes
> As an alternative, I could move this to a helper method

It's also not uncommon to have unstable logs and to use `toString` for logging. This practice alone cannot be used as a justification for not using `toString` for stable logging. Instead, we should stop using the practice.

----
<sup>1</sup> The more stable the logging, the more message-specific automation users will introduce to analyze them. When the amount of such automation reaches a critical point, we may end up having to support not only the API but also the arbitrary logging format we will have by that time. If this is what we are trying to achieve, then we are fine, but I would expect that we don't want this to happen. Some libraries intentionally violate properties that are not guaranteed just to reduce the chances of users relying on such properties, even though it is not necessary and the current state of the library manifests the property at runtime otherwise.

<sup>2</sup> I think, logging is what the method is supposed to be used. Do you think there are other more appropriate usages?

Reply:
The toString implementations that we typically provide are nearly all auto-generated, and useful only for debugging, not logging.  And worse, because toString is often called implicitly by `String.format`, you can't easily find usages of the method to see how it's being used. 

I would be happy to encapsulate this pretty much anywhere else than toString.

Reply:
> The toString implementations that we typically provide are nearly all auto-generated (and typically contain newlines)

I don't think this is the case, as looking through about 100 implementations in IDE with find filter "String toString()" shows otherwise - there are no explicit newline characters in the returned strings. IntelliJ Idea also does not insert such characters when it generates `toString`.

> intended for debugging, not logging

Logging exists for debugging purposes. But you are probably talking about the process of debugging via an interactive debugger in IDE, for example. This process is sensitive neither to the format of the `toString` result nor to its completeness (when the result is too long, one looks for the specific field of interest by means provided by a debugger). Given the two above statements, `toString` used for logging is equally useful for interactive debugging.

> because toString is often called implicitly by String.format, you can't easily find usages of the method to see where it's being called from

One don't necessary need to find usages, because the answer is supposed to be "it is called when we log messages".

Besides verbosity of the logging code, the problem with deconstructing objects for logging instead of using `toString` is that it significantly increases chances (comparing to when `toString` is used) of having different representations of the same object when logging the object from different places.

I presented all the arguments I can think of at the moment, which is one of the reasons I am approving the PR.

Reply:
Sorry, you are totally right about the newlines.  I was remembering incorrectly because of the way the toString methods are typically formatted with one field per line.  But still, something like this is not what I would want to log from a library like ours:

```
      "AcknowledgedBulkWriteResult{"
               + "insertedCount=" + insertedCount
               + ", matchedCount=" + matchedCount
               + ", removedCount=" + removedCount
               + ", modifiedCount=" + modifiedCount
               + ", upserts=" + upserts
               + '}'
```

Thank you for thinking about this as carefully as you have.  We should probably continue the more general discussion in a more interactive forum.